producer/kinesis.go

```diff
@@ -115,9 +115,101 @@ const (
 )
 
 type streamData struct {
-	content            *kinesis.PutRecordsInput
-	original           [][]*core.Message
+	content            []*kinesis.PutRecordsInput
+	original           [][][]*core.Message
+	streamName         string
 	lastRecordMessages int
+	lastRequestSize    int
+	recordMaxMessages  int
+	recordMaxSize      int
+	requestMaxSize     int
+	requestMaxRecords  int
+}
+
+func newStreamData(streamName string, recordMaxMessages int) *streamData {
+	return &streamData{
+		content: make([]*kinesis.PutRecordsInput, 0, 1),
+		original: make([][][]*core.Message, 0, 1),
+		streamName: streamName,
+		lastRecordMessages: 0,
+		lastRequestSize: 0,
+		recordMaxMessages: recordMaxMessages,
+		recordMaxSize: 1<<20, // 1 MB per record is the api limit
+		requestMaxSize: 5<<20, // 5 MB per request is the api limit
+		requestMaxRecords: 500, // 500 records per request is the api limit
+	}
+}
+
+func (sd *streamData) AddPutRecordsInput() *kinesis.PutRecordsInput {
+	input := &kinesis.PutRecordsInput{
+		Records:    make([]*kinesis.PutRecordsRequestEntry, 0, sd.requestMaxRecords),
+		StreamName: aws.String(sd.streamName),
+	}
+	sd.content = append(sd.content, input)
+	sd.original = append(sd.original, make([][]*core.Message, 0, sd.requestMaxRecords))
+	sd.lastRequestSize = 0
+	return input
+}
+
+func (sd *streamData) GetPutRecordsInput() (*kinesis.PutRecordsInput) {
+	if len(sd.content) == 0 {
+		return sd.AddPutRecordsInput()
+	} else {
+		return sd.content[len(sd.content)-1]
+	}
+}
+
+func (sd *streamData) AddRecord() *kinesis.PutRecordsRequestEntry {
+	input := sd.GetPutRecordsInput()
+	record := &kinesis.PutRecordsRequestEntry{
+		Data: make([]byte, 0, 0),
+	}
+	if len(input.Records) + 1 > sd.requestMaxRecords {
+		input = sd.AddPutRecordsInput()
+	}
+	input.Records = append(input.Records, record)
+	sd.original[len(sd.original)-1] = append(sd.original[len(sd.original)-1], make([]*core.Message, 0, sd.recordMaxMessages))
+	sd.lastRecordMessages = 0
+	return record
+}
+
+func (sd *streamData) GetRecord() *kinesis.PutRecordsRequestEntry {
+	input := sd.GetPutRecordsInput()
+	if len(input.Records) == 0 {
+		return sd.AddRecord()
+	} else {
+		return input.Records[len(input.Records)-1]
+	}
+}
+
+func (sd *streamData) AddMessage(delimiter []byte, data []byte, streamID core.MessageStreamID, msg *core.Message) {
+	record := sd.GetRecord()
+	if len(record.Data) > 0 {
```

Comment:
A benefit from encapsulating the request in an extra type is that you can store the size for each request in there, too. This will make the check for "am I too large" a simple
```go
func (sr streamRecord) OverLimt(data, delimiter []byte) {
    dataSize := len(data)
    if len(sr.records) > 0 {
        dataSize += len(delimiter)
    }
    return sr.currentSize + dataSize > sr.Limit
}

// ...
// func (sd *streamData) AddMessage ... {
    record := sd.GetCurrentRecord()
    if record.OverLimit(data,delimiter) {
        record = sd.AddNewRecord(/* ... */)
    }
    record.AddMessage(/* ... */)
```

This will make it a lot clearer to understand and a lot errorprone as you don't have to expose internal state.
E.g. it was very confusing for me why you check for `record.Data` twice though it wasn't directly visible where it is modified.